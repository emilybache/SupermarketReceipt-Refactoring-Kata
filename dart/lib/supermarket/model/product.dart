import 'package:supermarket_receipt/supermarket/model/product_unit.dart';

class Product {

  final String _name;
  final ProductUnit _unit;

  Product(this._name, this._unit);

  String getName() => _name;

  ProductUnit getUnit() => _unit;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Product &&
          runtimeType == other.runtimeType &&
          _name == other._name &&
          _unit == other._unit;

  @override
  int get hashCode => _name.hashCode ^ _unit.hashCode;
}