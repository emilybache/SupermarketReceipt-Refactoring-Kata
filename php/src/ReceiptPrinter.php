<?php

declare(strict_types=1);

namespace Supermarket;

use Supermarket\Model\{
    Discount,
    ProductUnit,
    Receipt,
    ReceiptItem
};

class ReceiptPrinter
{
    /**
     * @var int
     */
    private $columns;

    public function __construct(int $columns = 40)
    {
        $this->columns = $columns;
    }

    public function printReceipt(Receipt $receipt): string
    {
        $result = '';
        foreach ($receipt->getItems() as $item) {
            $itemPresentation = $this->presentReceiptItem($item);
            $result .= $itemPresentation;
        }

        foreach ($receipt->getDiscounts() as $discount) {
            $discountPresentation = $this->presentDiscount($discount);
            $result .= $discountPresentation;
        }

        $result .= "\n";
        $result .= $this->presentTotal($receipt);
        return $result;
    }

    protected function presentReceiptItem(ReceiptItem $item): string
    {
        $price = self::presentPrice($item->getTotalPrice());
        $name = $item->getProduct()->getName();

        $line = $this->formatLineWithWhitespace($name, $price) . "\n";

        if ($item->getQuantity() !== 1.0) {
            $line .= '  ' . self::presentPrice($item->getPrice()) . ' * ' . self::presentQuantity($item) . "\n";
        }
        return $line;
    }

    protected function presentDiscount(Discount $discount): string
    {
        $name = "{$discount->getDescription()}({$discount->getProduct()->getName()})";
        $value = self::presentPrice($discount->getDiscountAmount());

        return $this->formatLineWithWhitespace($name, $value) . "\n";
    }

    protected function presentTotal(Receipt $receipt): string
    {
        $name = 'Total: ';
        $value = self::presentPrice($receipt->getTotalPrice());
        return $this->formatLineWithWhitespace($name, $value);
    }

    protected function formatLineWithWhitespace(string $name, string $value): string
    {
        $whitespaceSize = $this->columns - strlen($name) - strlen($value);
        return $name . str_repeat(' ', $whitespaceSize) . $value;
    }

    protected static function presentPrice(float $price): string
    {
        return sprintf('%.2F', $price);
    }

    private static function presentQuantity(ReceiptItem $item): string
    {
        return $item->getProduct()->getUnit()->equals(ProductUnit::EACH()) ?
            sprintf('%x', $item->getQuantity()) :
            sprintf('%.3F', $item->getQuantity());
    }
}
