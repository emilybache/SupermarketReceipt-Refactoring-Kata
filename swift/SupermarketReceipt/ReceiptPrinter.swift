public class ReceiptPrinter {

    private var columns: Int = 40

    public init(columns: Int) {
        self.columns = columns
    }

    public func printReceipt(receipt: Receipt) -> String {
        var result = ""
        for item in receipt.items {
            var price = String(format: "%.2f", item.totalPrice)
            var quantity = ReceiptPrinter.presentQuantity(item: item)
            var name = item.product.name
            var unitPrice = String(format :"%.2f", item.price)

            var whitespaceSize = self.columns - name.count - price.count
            var line = name + ReceiptPrinter.getWhitespace(whitespaceSize: whitespaceSize) + price + "\n"

            if (item.quantity != 1) {
                line += "  " + unitPrice + " * " + quantity + "\n"
            }
            result.append(line)
        }
        for discount in receipt.discounts {
            var productPresentation = discount.product.name
            var pricePresentation = String(format: "%.2f", discount.discountAmount)
            var description = discount.description
            result.append(description)
            result.append("(")
            result.append(productPresentation)
            result.append(")")
            result.append(ReceiptPrinter.getWhitespace(whitespaceSize: self.columns - 3 - productPresentation.count - description.count - pricePresentation.count))
            result.append("-")
            result.append(pricePresentation)
            result.append("\n")
        }
        result.append("\n")
        var pricePresentation = String(format: "%.2f", Double(receipt.getTotalPrice()))
        var total = "Total: "
        var whitespace = ReceiptPrinter.getWhitespace(whitespaceSize: self.columns - total.count - pricePresentation.count)
        result.append(total)
        result.append(whitespace)
        result.append(pricePresentation)
        return result
    }

    private static func presentQuantity(item: ReceiptItem ) -> String {
        return ProductUnit.Each == item.product.unit
            ? String(format: "%x", Int(item.quantity))
            : String(format: "%.3f", item.quantity)
    }

    private static func getWhitespace(whitespaceSize: Int) -> String {
        var whitespace = ""
        for i in 0..<whitespaceSize {
            whitespace.append(" ")
        }
        return whitespace
    }
}
