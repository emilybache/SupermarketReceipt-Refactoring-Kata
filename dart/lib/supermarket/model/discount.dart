import 'package:supermarket_receipt/supermarket/model/product.dart';

class Discount {

  final Product _product;
  final String _description;
  final double _discountAmount;

  Discount(this._product, this._description, this._discountAmount);

  String getDescription() => _description;

  double getDiscountAmount() => _discountAmount;

  Product getProduct() => _product;
}