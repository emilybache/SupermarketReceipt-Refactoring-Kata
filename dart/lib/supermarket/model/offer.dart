import 'package:supermarket_receipt/supermarket/model/product.dart';
import 'package:supermarket_receipt/supermarket/model/special_offer_type.dart';


class Offer {

  final SpecialOfferType offerType;
  final Product _product;
  final double argument;

  Offer(this.offerType, this._product, this.argument);

  Product getProduct() => _product;
}