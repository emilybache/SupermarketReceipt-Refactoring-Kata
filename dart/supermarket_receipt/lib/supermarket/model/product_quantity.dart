import 'package:supermarket_receipt/supermarket/model/product.dart';

class ProductQuantity {

  final Product _product;
  final double _quantity;

  ProductQuantity(this._product, this._quantity);

  Product getProduct() => _product;

  double getQuantity() => _quantity;
}