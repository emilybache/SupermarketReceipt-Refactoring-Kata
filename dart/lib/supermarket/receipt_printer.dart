import 'package:supermarket_receipt/supermarket/model/product_unit.dart';
import 'package:supermarket_receipt/supermarket/model/receipt.dart';
import 'package:supermarket_receipt/supermarket/model/receipt_item.dart';
import 'package:supermarket_receipt/supermarket/model/discount.dart';

class ReceiptPrinter {
  
  final int _columns;

  ReceiptPrinter() : this._columns = 40;

  ReceiptPrinter.withColumns(this._columns);

  String printReceipt(Receipt receipt) {
    StringBuffer result = StringBuffer();
    for (ReceiptItem item in receipt.getItems()) {
      String receiptItem = presentReceiptItem(item);
      result.write(receiptItem);
    }
    for (Discount discount in receipt.getDiscounts()) {
      String discountPresentation = presentDiscount(discount);
      result.write(discountPresentation);
    }

    result.write('\n');
    result.write(presentTotal(receipt));
    return result.toString();
  }

  String presentReceiptItem(ReceiptItem item) {
    String totalPricePresentation = presentPrice(item.getTotalPrice());
    String name = item.getProduct().getName();

    String line = formatLineWithWhitespace(name, totalPricePresentation);

    if (item.getQuantity() != 1) {
      line += '  ' + presentPrice(item.getPrice()) + ' * ' + presentQuantity(item) + '\n';
    }
    return line;
  }

  String presentDiscount(Discount discount) {
    String name = discount.getDescription() + ' (' + discount.getProduct().getName() + ')';
    String value = presentPrice(discount.getDiscountAmount());

    return formatLineWithWhitespace(name, value);
  }

  String presentTotal(Receipt receipt) {
    String name = 'Total: ';
    String value = presentPrice(receipt.getTotalPrice());
    return formatLineWithWhitespace(name, value);
  }

  String formatLineWithWhitespace(String name, String value) {
    StringBuffer line = StringBuffer();
    line.write(name);
    int whitespaceSize = _columns - name.length - value.length;
    for (int i = 0; i < whitespaceSize; i++) {
      line.write(' ');
    }
    line.write(value);
    line.write('\n');
    return line.toString();
  }

  String presentPrice(double price) {
    return price.toStringAsFixed(2);
  }

  String presentQuantity(ReceiptItem item) {
    if (item.getProduct().getUnit() == ProductUnit.EACH) {
      return item.getQuantity().toInt().toString();
    } else {
      return item.getQuantity().toStringAsFixed(3);
    }
  }
}