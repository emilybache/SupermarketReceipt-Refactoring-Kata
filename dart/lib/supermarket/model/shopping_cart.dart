import 'package:supermarket_receipt/supermarket/model/discount.dart';
import 'package:supermarket_receipt/supermarket/model/offer.dart';
import 'package:supermarket_receipt/supermarket/model/product.dart';
import 'package:supermarket_receipt/supermarket/model/product_quantity.dart';
import 'package:supermarket_receipt/supermarket/model/receipt.dart';
import 'package:supermarket_receipt/supermarket/model/special_offer_type.dart';
import 'package:supermarket_receipt/supermarket/model/supermarket_catalog.dart';

class ShoppingCart {

  final List<ProductQuantity> _items = [];
  final Map<Product, double> _productQuantities = {};

  void addItem(Product product) {
    addItemQuantity(product, 1.0);
  }

  void addItemQuantity(Product product, double quantity) {
    _items.add(ProductQuantity(product, quantity));
    if (_productQuantities.containsKey(product)) {
      _productQuantities[product] = _productQuantities[product]! + quantity;
    } else {
      _productQuantities[product] = quantity;
    }
  }

  List<ProductQuantity> getItems() {
    return _items;
  }

  Map<Product, double> getProductQuantities() {
    return _productQuantities;
  }

  void handleOffers(Receipt receipt, Map<Product, Offer> offers, SupermarketCatalog catalog) {
    for (Product p in _productQuantities.keys) {
      double quantity = _productQuantities[p]!;
      if (offers.containsKey(p)) {
        Offer offer = offers[p]!;
        double unitPrice = catalog.getUnitPrice(p);
        int quantityAsInt = quantity.toInt();
        Discount? discount = null;
        int x = 1;
        if (offer.offerType == SpecialOfferType.THREE_FOR_TWO) {
          x = 3;
        }
        else if (offer.offerType == SpecialOfferType.TWO_FOR_AMOUNT) {
          x = 2;
          if (quantityAsInt >= 2) {
            double total = offer.argument * (quantityAsInt ~/ x) + quantityAsInt % 2 * unitPrice;
            double discountN = unitPrice * quantity - total;
            discount = Discount(p, "2 for " + offer.argument.toString(), -discountN);
          }
        }
        else if (offer.offerType == SpecialOfferType.FIVE_FOR_AMOUNT) {
          x = 5;
        }
        int numberOfXs = quantityAsInt ~/ x;
        if (offer.offerType == SpecialOfferType.THREE_FOR_TWO && quantityAsInt > 2) {
          double discountAmount = quantity * unitPrice - ((numberOfXs * 2 * unitPrice) + quantityAsInt % 3 * unitPrice);
          discount = Discount(p, "3 for 2", -discountAmount);
        }
        if (offer.offerType == SpecialOfferType.TEN_PERCENT_DISCOUNT) {
          discount = Discount(p, offer.argument.toString() + "% off", -quantity * unitPrice * offer.argument / 100.0);
        }
        if (offer.offerType == SpecialOfferType.FIVE_FOR_AMOUNT && quantityAsInt >= 5) {
          double discountTotal = unitPrice * quantity - (offer.argument * numberOfXs + quantityAsInt % 5 * unitPrice);
          discount = Discount(p, x.toString() + " for " + offer.argument.toString(), -discountTotal);
        }
        if (discount != null) receipt.addDiscount(discount);
      }
    }
  }
}