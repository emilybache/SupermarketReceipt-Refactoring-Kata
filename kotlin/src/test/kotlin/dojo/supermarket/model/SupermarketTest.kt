package dojo.supermarket.model

import org.approvaltests.Approvals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import supermarket.ReceiptPrinter
import supermarket.model.*

class SupermarketTest {
    private lateinit var catalog: SupermarketCatalog
    private lateinit var teller: Teller
    private lateinit var theCart: ShoppingCart
    private lateinit var toothbrush: Product
    private lateinit var rice: Product
    private lateinit var apples: Product
    private lateinit var cherryTomatoes: Product
    
    @BeforeEach
    fun setUp() {
        catalog = FakeCatalog()
        teller = Teller(catalog)
        theCart = ShoppingCart()

        toothbrush = Product("toothbrush", ProductUnit.Each)
        catalog.addProduct(toothbrush, 0.99)
        rice = Product("rice", ProductUnit.Each)
        catalog.addProduct(rice, 2.99)
        apples = Product("apples", ProductUnit.Kilo)
        catalog.addProduct(apples, 1.99)
        cherryTomatoes = Product("cherry tomato box", ProductUnit.Each)
        catalog.addProduct(cherryTomatoes, 0.69)

    }

    @Test
    fun an_empty_shopping_cart_should_cost_nothing() {
        val receipt = teller.checksOutArticlesFrom(theCart)
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun one_normal_item() {
        theCart.addItem(toothbrush)
        val receipt = teller.checksOutArticlesFrom(theCart)
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun two_normal_items() {
        theCart.addItem(toothbrush)
        theCart.addItem(rice)
        val receipt = teller.checksOutArticlesFrom(theCart)
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun buy_two_get_one_free() {
        theCart.addItem(toothbrush)
        theCart.addItem(toothbrush)
        theCart.addItem(toothbrush)
        teller.addSpecialOffer(SpecialOfferType.ThreeForTwo, toothbrush, catalog.getUnitPrice(toothbrush))
        val receipt = teller.checksOutArticlesFrom(theCart)
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun buy_five_get_one_free() {
        theCart.addItem(toothbrush)
        theCart.addItem(toothbrush)
        theCart.addItem(toothbrush)
        theCart.addItem(toothbrush)
        theCart.addItem(toothbrush)
        teller.addSpecialOffer(SpecialOfferType.ThreeForTwo, toothbrush, catalog.getUnitPrice(toothbrush))
        val receipt = teller.checksOutArticlesFrom(theCart)
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun loose_weight_product() {
        theCart.addItemQuantity(apples, .5)
        val receipt = teller.checksOutArticlesFrom(theCart)
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun percent_discount() {
        theCart.addItem(rice)
        teller.addSpecialOffer(SpecialOfferType.TenPercentDiscount, rice, 10.0)
        val receipt = teller.checksOutArticlesFrom(theCart)
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun xForY_discount() {
        theCart.addItem(cherryTomatoes)
        theCart.addItem(cherryTomatoes)
        teller.addSpecialOffer(SpecialOfferType.TwoForAmount, cherryTomatoes, .99)
        val receipt = teller.checksOutArticlesFrom(theCart)
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun FiveForY_discount() {
        theCart.addItemQuantity(apples, 5.0)
        teller.addSpecialOffer(SpecialOfferType.FiveForAmount, apples, 6.99)
        val receipt = teller.checksOutArticlesFrom(theCart)
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun FiveForY_discount_withSix() {
        theCart.addItemQuantity(apples, 6.0)
        teller.addSpecialOffer(SpecialOfferType.FiveForAmount, apples, 6.99)
        val receipt = teller.checksOutArticlesFrom(theCart)
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun FiveForY_discount_withSixteen() {
        theCart.addItemQuantity(apples, 16.0)
        teller.addSpecialOffer(SpecialOfferType.FiveForAmount, apples, 6.99)
        val receipt = teller.checksOutArticlesFrom(theCart)
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }

    @Test
    fun FiveForY_discount_withFour() {
        theCart.addItemQuantity(apples, 4.0)
        teller.addSpecialOffer(SpecialOfferType.FiveForAmount, apples, 6.99)
        val receipt = teller.checksOutArticlesFrom(theCart)
        Approvals.verify(ReceiptPrinter(40).printReceipt(receipt))
    }
}
