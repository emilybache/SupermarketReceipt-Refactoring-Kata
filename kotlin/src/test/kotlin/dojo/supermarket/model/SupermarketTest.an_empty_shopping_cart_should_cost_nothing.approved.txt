
Total:                              0.00
